tokio/src/sync/watch.rs

```diff
@@ -551,14 +507,13 @@ impl<T> Receiver<T> {
     /// }
     /// ```
     pub fn has_changed(&self) -> Result<bool, error::RecvError> {
-        // Load the version from the state
-        let state = self.shared.state.load();
-        if state.is_closed() {
+        if self.shared.is_sender_closed() {
             // The sender has dropped.
             return Err(error::RecvError(()));
         }
-        let new_version = state.version();
 
+        // Load the version from the state
+        let new_version = self.shared.version();
         Ok(self.version != new_version)
```

Comment:
Since you've split the closed bit out from the atomic, this introduces extra possibilities for race conditions causing incorrect behavior. I would need some time to think about whether this is still correct.